core/src/main/java/com/google/cloud/sql/core/Refresher.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/** Handles periodic refresh operations for an instance. */
+class Refresher<T extends ExpiringData> {
```

Comment:
Does this need to be generic? We'll only ever be refreshing "InstanceData" aka "ConnectionInfo".

Reply:
For now, yes. We want to copy/paste this into AlloyDB unchanged.  Once we're using the same name for ConnectionInfo in both projects, we can eliminate the type parameter. 

Also, in the future I can see us using this for other things too - like refreshing just the IAM Authn credentials for an implementation of HikariPool that supports "direct path iam authn" connections.

Reply:
Copy/paste + search and replace is cheap. I think we want to avoid introducing this abstraction until we need it.